sys/net/crosslayer/ng_pktbuf_default/ng_pktbuf_default.c

```diff
@@ -0,0 +1,406 @@
+     *  - the head of _unused_t list is _first_unused
+     *  - if _unused_t list is empty the packet buffer is full and _first_unused is NULL
+     *  - forall ptr_in _unused_t list: &_pktbuf[0] < ptr < &_pktbuf[NG_PKTBUF_SIZE]
+     *  - forall ptr in _unused_t list: ptr->next == NULL || ptr < ptr->next
+     *  - forall ptr in _unused_t list: (ptr->next != NULL && ptr->size <= (ptr->next - ptr)) ||
+     *                                  (ptr->next == NULL && ptr->size == (NG_PKTBUF_SIZE - (ptr - &_pktbuf[0])))
+     */
+
+    while (ptr) {
+        if (&_pktbuf[0] >= (uint8_t *)ptr && (uint8_t *)ptr >= &_pktbuf[NG_PKTBUF_SIZE]) {
+            return false;
+        }
+        if ((ptr->next != NULL) && (ptr >= ptr->next)) {
+            return false;
+        }
+        if (((ptr->next == NULL) || (ptr->size > (size_t)((uint8_t *)(ptr->next) - (uint8_t *)ptr))) &&
+            ((ptr->next != NULL) || (ptr->size != (size_t)(NG_PKTBUF_SIZE - ((uint8_t *)ptr - &_pktbuf[0]))))) {
+            return false;
+        }
+        ptr = ptr->next;
+    }
+
+    return true;
+}
+#endif
+
+static ng_pktsnip_t *_create_snip(ng_pktsnip_t *next, void *data, size_t size,
+                                  ng_nettype_t type)
+{
+    ng_pktsnip_t *pkt = _pktbuf_alloc(sizeof(ng_pktsnip_t));
+    void *_data;
+    if (pkt == NULL) {
+        DEBUG("pktbuf: error allocating new packet snip\n");
+        return NULL;
+    }
+    _data = _pktbuf_alloc(size);
+    if (_data == NULL) {
+        DEBUG("pktbuf: error allocating data for new packet snip\n");
+        _pktbuf_free(pkt, sizeof(ng_pktsnip_t));
+        return NULL;
+    }
+    pkt->next = next;
+    pkt->size = size;
+    pkt->data = _data;
+    pkt->type = type;
+    pkt->users = 1;
+    if (data != NULL) {
+        memcpy(_data, data, size);
+    }
+    return pkt;
+}
+
+static void *_pktbuf_alloc(size_t size)
+{
+    _unused_t *prev = NULL, *ptr = _first_unused;
+    size = (size < sizeof(_unused_t)) ? _align(sizeof(_unused_t)) : _align(size);
+    while (ptr && size > ptr->size) {
+        prev = ptr;
+        ptr = ptr->next;
+    }
+    if (ptr == NULL) {
+        DEBUG("pktbuf: no space left in packet buffer\n");
+        return NULL;
+    }
+    if (sizeof(_unused_t) > (ptr->size - size)) {
+        if (prev == NULL) { /* ptr was _first_unused */
+            _first_unused = ptr->next;
+        }
+        else {
+            prev->next = ptr->next;
+        }
+    }
+    else {
+        _unused_t *new = (_unused_t *)(((uint8_t *)ptr) + size);
+        if (prev == NULL) { /* ptr was _first_unused */
+            _first_unused = new;
+        }
+        else {
+            prev->next = new;
+        }
+        new->next = ptr->next;
+        new->size = ptr->size - size;
+    }
+    return (void *)ptr;
+}
+
+static inline bool _too_small_hole(_unused_t *a, _unused_t *b)
+{
+    return sizeof(_unused_t) > (size_t)(((uint8_t *)b) - (((uint8_t *)a) + a->size));
```

Comment:
is it really necessary to add `a->size` here? I assumed that the address of `b` is already at least `a->size` bytes next to `a`